### PR TITLE
WorkMail Upstream Gateway Filter

### PR DESCRIPTION
**Pull request by zjt**
*Description of changes:*
This is a new WorkMail Lambda solution that matches a pattern in the message header of inbound mail to filter messages to Junk E-Mail. It is useful for organizations who use upstream SMTP email security gateways to detect spam and other unwanted email.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workmail-upstream-gateway-filter/src/app.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    msg_id = email_summary['messageId']
+    raw_msg = workmail.get_raw_message_content(messageId=msg_id)
+    parsed_msg = email.message_from_bytes(raw_msg['messageContent'].read())
+    filter_header_value = parsed_msg.get(filter_header_name)
+    logger.info(filter_header_value)
+
+    flow_direction = email_summary['flowDirection']
+    
+    if flow_direction == 'INBOUND':
+        if regexp.search(filter_header_value):
```
Comment (sagariiit) [inline workmail-upstream-gateway-filter/src/app.py:104]: If the header is not found then search fails with `TypeError`, so handle missing header properly by returning `DEFAULT`?

File: workmail-upstream-gateway-filter/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# Amazon WorkMail Upstream Gateway Filter
+
+This application enables you to filter messages to Junk E-Mail for multiple recipeints based on the value of an email header added by an upstream email security gateway.
```
Comment (sagariiit) [inline workmail-upstream-gateway-filter/README.md:3]: Once we have your BlogPost live, then may be we can link it here?

File: workmail-upstream-gateway-filter/src/app.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            },
+            'recipients': list of strings,                          # Optional. Indicates list of recipients for which this action applies
+            'default': boolean                                      # Optional. Indicates whether this action applies to all recipients
+          }
+        ]}
+
+    """
+    
+    # get the environment variables containing the header name and regex to match
+    filter_header_name = os.getenv('FILTER_HEADER_NAME')
```
[the remark this answers is not in the document]
Comment (zjt) [reply workmail-upstream-gateway-filter/src/app.py:89]: I'll do the same for the regex too

File: workmail-upstream-gateway-filter/README.md (hunk 2 above)
Comment (zjt) [reply workmail-upstream-gateway-filter/README.md:3]: Yes, that is what I am thinking too

**Comment by zjt** [comment]
OK, I believe I addressed all of the suggestions in my latest push to the master branch on zjt/amazon-workmail-lambda-templates.

**Review by sagariiit — APPROVED** [review]

**Pull request merged**